### ISSUE TITLE
wmllint thinks the `id=` key for the `[leader_goal]` tag is the same as the `id=` key for units

### ISSUE THREAD

**Issue by cooljeanius** [wesnoth/wesnoth#6395]
In cooljeanius/A_New_Order@388a300 I added `id=` keys to the `[leader_goal]` tags in one of the scenarios, in case I might want to try removing some of them via WML under certain conditions later. Unfortunately, this leads to new `wmllint` warnings:
```
"/Users/ericgallager/Library/Containers/org.wesnoth.Wesnoth/Data/Library/Application Support/Wesnoth_1.16/data/add-ons/A_New_Order/scenarios/07_Ally_From_the_Past.cfg", line 149: unknown 'report_to_Bor_Cryne_1' referred to by id
"/Users/ericgallager/Library/Containers/org.wesnoth.Wesnoth/Data/Library/Application Support/Wesnoth_1.16/data/add-ons/A_New_Order/scenarios/07_Ally_From_the_Past.cfg", line 182: unknown 'report_to_Bor_Cryne_2' referred to by id
"/Users/ericgallager/Library/Containers/org.wesnoth.Wesnoth/Data/Library/Application Support/Wesnoth_1.16/data/add-ons/A_New_Order/scenarios/07_Ally_From_the_Past.cfg", line 195: unknown 'report_to_Bor_Cryne_3' referred to by id
"/Users/ericgallager/Library/Containers/org.wesnoth.Wesnoth/Data/Library/Application Support/Wesnoth_1.16/data/add-ons/A_New_Order/scenarios/07_Ally_From_the_Past.cfg", line 208: unknown 'report_to_Bor_Cryne_4' referred to by id
"/Users/ericgallager/Library/Containers/org.wesnoth.Wesnoth/Data/Library/Application Support/Wesnoth_1.16/data/add-ons/A_New_Order/scenarios/07_Ally_From_the_Past.cfg", line 221: unknown 'report_to_Bor_Cryne_5' referred to by id
"/Users/ericgallager/Library/Containers/org.wesnoth.Wesnoth/Data/Library/Application Support/Wesnoth_1.16/data/add-ons/A_New_Order/scenarios/07_Ally_From_the_Past.cfg", line 238: unknown 'report_to_Bor_Cryne_6' referred to by id
```
I asked @Elvish-Hunter about this and apparently it is a bug. So, I'm opening this issue for it.

**Comment by CelticMinstrel** [comment]
I'm going to note that _any_ ID nested anywhere within an `[ai]` tag should be ignored, not just `[leader_goal]id=`.

**Comment by cooljeanius** [comment]
> I'm going to note that _any_ ID nested anywhere within an `[ai]` tag (also possibly `[modify_ai]` and `[micro_ai]`) should be ignored, not just `[leader_goal]id=`.

oh yeah those too; I just remembered that I have actually seen the MicroAI one myself in this same add-on...

**Comment by CelticMinstrel** [comment]
Yeah, and if I recall correctly, there are a _lot_ of tags in `[ai]` that can have an ID, like `[aggression]` for example. Basically every aspect can be a tag. Then of course there's `[goal]`, `[engine]`, and `[stage]`… but probably easiest to just ignore anything that's nested under an `[ai]` tag. It's not like there's any possibility of units being nested under an `[ai]` tag. Although _references_ to units are certainly possible… there are quite a few tags that take unit filters.

**Comment by cooljeanius** [comment]
> Yeah, and if I recall correctly, there are a _lot_ of tags in `[ai]` that can have an ID, like `[aggression]` for example. Basically every aspect can be a tag. Then of course there's `[goal]`, `[engine]`, and `[stage]`… but probably easiest to just ignore anything that's nested under an `[ai]` tag. It's not like there's any possibility of units being nested under an `[ai]` tag. Although _references_ to units are certainly possible… there are quite a few tags that take unit filters.

`[aspect]` and `[facet]` too, as I recently learned, but luckily `wmllint` hasn't complained about any of their `id=` keys for me yet...

**Comment by CelticMinstrel** [comment]
It's just kind of complicated to handle this in wmllint, since it doesn't actually parse the WML. I _think_ you could get it almost perfect by turning off id detection when entering an `[ai]` tag, and turning it back on when you enter certain other tags, but it's getting that list of "certain other tags" that might be the tricky part. For aspects that could contain a unit ID, there's avoid and attacks… I think those are the only ones? So for avoid the tag could be either `[avoid]` or `[value]`, and for attacks the tags are `[filter_own]` and `[filter_enemy]`. There's also leader_id in `[recruit]` which complicates things. For goals, there's either the short syntax (where the tag is the type of goal) or the long syntax (where the tag is `[criteria]`). Then there's the fact that you can nest a `[micro_ai]` tag inside `[ai]`, which could contain unit IDs, usually in tags of the form `[filter*]`.

So basically, the AI section is quite complicated, and getting wmllint to be perfect in this respect may not be worth the effort.

I think a schema validation tool could actually do a pretty good job, though. I'm not entirely sure how this would look, and it may require some new feature in the schema, but it could probably do a much better job in most cases. Obviously that doesn't help much here though…

**Comment by Wedge009** [comment]
Just noting that this is an issue that arises with Winds of Fate campaign in current master.

**Comment by cooljeanius** [comment]
Seen also in: https://github.com/cooljeanius/Son_Of_The_Black_Eye_Easy_Edition/actions/runs/7363882244/job/20043755361

**Comment by stevecotton** [comment]
There's a giant list of tags that cause wmllint to ignore `id=`, it's around line 2560 of wmllint (as of the 1.17.26 tag). This should be fixed by adding `[leader_goal]` there.

**Comment by CelticMinstrel** [comment]
Probably IDREF types in the schema would do a better job of this kind of checking. If XML schemas can do it, presumably we could find a way to code it into WML schemas too.
